pep-9999.rst

```diff
@@ -0,0 +1,736 @@
+
+Intelligent indexing of structured data
+---------------------------------------
+
+Literals can be used to "intelligently index" into structured types like
+tuples, NamedTuple, and classes. (Note: this is not an exhaustive list).
+
+For example, type checkers should infer the correct value type when
+indexing into a tuple using an int key that corresponds a valid index::
+
+   a: Literal[0] = 0
+   b: Literal[5] = 5
+
+   some_tuple: Tuple[int, str, List[bool]] = (3, "abc", [True, False])
+   reveal_type(some_tuple[a])   # Revealed type is 'int'
+   some_tuple[b]                # Error: 5 is not a valid index into the tuple
+
+We expect similar behavior when using functions like getattr::
+
+   class Test:
+       def __init__(self, param: int) -> None:
+           self.myfield = param
+       
+       def mymethod(self, val: int) -> str: ...
+       
+   a: Literal["myfield"]  = "myfield"
+   b: Literal["mymethod"] = "mymethod"
+   c: Literal["blah"]     = "blah"
+
+   t = Test()
+   reveal_type(getattr(t, a))  # Revealed type is 'int'
+   reveal_type(getattr(t, b))  # Revealed type is 'Callable[[int], str]'
+   getattr(t, c)               # Error: No attribute named 'blah' in Test
+
+Interactions with overloads
+---------------------------
+
+Literal types and overloads do not need to interact in  a special
+way: the existing rules work fine.
+
+However, one important use case type checkers must take care to
+support is the ability a *fallback* when the user is not using literal
+types. For example, consider ``open``::
+
+   _PathType = Union[str, bytes, int]
+
+   @overload
+   def open(path: _PathType, 
+            mode: Literal["r", "w", "a", "x", "r+", "w+", "a+", "x+"],
+            ) -> IO[Text]: ...
+   @overload
+   def open(path: _PathType, 
+            mode: Literal["rb", "wb", "ab", "xb", "r+b", "w+b", "a+b", "x+b"],
+            ) -> IO[bytes]: ...
+
+   # Fallback overload for when the user isn't using literal types
+   @overload
+   def open(path: _PathType, mode: str) -> IO[Any]: ...
+
+If we change the signature of ``open`` to use just the first two overloads,
+we would break any code that does not pass in a literal string expression.
+For example, code like this would be broken::
+
+   mode: str = pick_file_mode(...)
+   with open(path, mode) as f:
+       # f should continue to be of type IO[Any] here
+
+A little more broadly: we propose adding a policy to typeshed that
+mandates that whenever we add literal types to some existing API, we also
+always include a fallback overload to maintain backwards-compatibility.
+
+Interactions with generics
+--------------------------
+
+Types like ``Literal[3]`` are meant to be just plain old subclasses of
+``int``. This means you can use types like ``Literal[3]`` anywhere
+you could use normal types, such as with generics.
+
+This means that it is legal to parameterize generic functions or
+classes using Literal types::
```

Comment:
I think this is potentially problematic re: backwards compatibility.
If we allow having typevariables bound to literals, then we run the risk of having variance issues crop up.  Avoiding these requires a degree of lookahead that seems like a big burden to push onto type checkers.
I think it might be prudent to ban them being propagated through normal typevariables, and instead wait for the IntVar proposal or something like that to specifically take them.
Short of that, perhaps we could only allow them in type variables like the ones in your example, upper bounded by the corresponding type?

Reply:
Sorry, I'm not quite sure what you mean -- could you perhaps provide an example of a variance or lookahead situation you're worried about?

If I had to guess, I'd imagine you're perhaps worrying about how to handle examples like this one?

    T = TypeVar('T')
    class MyContainer(Generic[T]): pass
    x = 5

    # Is y of type MyContainer[int] or MyContainer[Literal[5]]?
    # We might need to see how y is used to know for certain?
    y = MyContainer(x)

I think it would be reasonable if Pyre decided against doing anything complicated here, in the same way that mypy doesn't try doing anything complicated w.r.t. inferring the type of unannotated variables. For example, it would be fine if you defaulted to having `MyContainer(...)` be of type `MyContainer[int]` in this case -- or maybe if you decided to only try the lookahead thing you described if the TypeVar had an upper bound.

I think Mypy would basically do something similar anyways, though that's really more a consequence of how we chose to handle variable assignments.

Anyways, while writing this section, the only thing I meant to require is that it should be possible to *parameterize* generics with Literal types -- e.g. write type hints that look like `MyContainer[Literal[5]]`, or explicitly set the type by doing something like:

    y = MyContainer[Literal[5]](5)

And once you do/once the typevars are bound, the type system should behave exactly as one might expect.

But I don't think it makes sense to prescribe any particular strategy or restriction on how Literals are propagated through TypeVars or how to infer what gets bound to a TypeVar in this PEP, in the same way we decided it didn't make sense to prescribe a particular strategy on how to handle type inference with variable assignment and the like.

(Let me know if I totally misunderstood your concern!)

---

FWIW, the motivation behind allowing Literals and generics to interact in this way is that we wanted to make Literal types behave as closely as possible to existing types -- the underlying story this PEP is painting is that `Literal[1]` is "just" a subtype of `int`. The concern is that users would find it confusing if we disallowed Literals from binding to TypeVars: it would feel like a weird special case.

I think this concern would hold even if we added IntVars. Then, I'd imagine users would start to wonder why they can't parameterize generics with Literal strings or Literal enum values.

Also, not sure if this helps, but when we were implementing this in mypy, we found we were able to directly apply the bulk of our generics and type inference related code to handle Literals with minimal special-casing. We treat Literals just like any other subtype, and only add in one special case to handle which "default type" gets picked when we encounter an expression with no real constraints on what type it could be.

Of course, I'm not at all familiar with how Pyre is implemented, so maybe it's not possible for you to adopt this strategy.